runner/app/live/streamer/process.py

```diff
@@ -117,91 +119,108 @@ def process_loop(self):
+            if params:
                 try:
                     with log_timing(f"PipelineProcess: Pipeline loading with default params due to error with params: {params}"):
                         pipeline = load_pipeline(self.pipeline_name)
+                        await pipeline.set_params()
+                        return pipeline
                 except Exception as e:
-                    report_error(f"Error loading pipeline with default params: {e}")
+                    self._report_error(f"Error loading pipeline with default params: {e}")
                     raise
 
-            while not self.is_done():
-                while not self.param_update_queue.empty():
-                    params = self.param_update_queue.get_nowait()
-                    try:
-                        logging.info(f"PipelineProcess: Processing parameter update from queue: {params}")
-                        if _handle_logging_params(params):
-                            logging.info(f"PipelineProcess: Updating pipeline parameters")
-                            pipeline.update_params(**params)
-                            logging.info(f"PipelineProcess: Successfully applied params to pipeline: {params}")
-                    except Exception as e:
-                        error_msg = f"Error updating params: {str(e)}"
-                        logging.error(error_msg, exc_info=True)
-                        report_error(error_msg)
-
-                try:
-                    input_frame = self.input_queue.get(timeout=0.1)
-                except queue.Empty:
-                    continue
+    async def _run_pipeline_loops(self):
+        pipeline = await self._initialize_pipeline()
+        await pipeline.warm_video()
+        input_task = asyncio.create_task(self._input_loop(pipeline))
+        output_task = asyncio.create_task(self._output_loop(pipeline))
+        param_task = asyncio.create_task(self._param_update_loop(pipeline))
 
-                try:
-                    if isinstance(input_frame, VideoFrame):
-                        input_frame.log_timestamps["pre_process_frame"] = time.time()
-                        output_image = pipeline.process_frame(input_frame.image)
-                        input_frame.log_timestamps["post_process_frame"] = time.time()
-                        output_frame = VideoOutput(input_frame.replace_image(output_image))
-                        self.output_queue.put(output_frame)
-                    elif isinstance(input_frame, AudioFrame):
-                        self.output_queue.put(AudioOutput([input_frame]))
-                        # TODO wire in a proper pipeline here
-                    else:
-                        report_error(f"Unsupported input frame type {type(input_frame)}")
-                except Exception as e:
-                    report_error(f"Error processing frame: {e}")
+        try:
+            await asyncio.gather(input_task, output_task, param_task)
```

Comment:
We should stop the whole thing if any of these tasks, otherwise we will continue running even with only part of the pipeline working, which is invalid. You can use `asyncio.wait` instead [like the streamer](https://github.com/livepeer/ai-worker/blob/29408e443488107eb9a33af06f01e6b0bbe46564/runner/app/live/streamer/streamer.py#L75)

Then on exceptions we should make sure we cancel the other tasks (or maybe just let them be, we're gonna exit the process anyway)

Reply:
Updated with asyncio.wait https://github.com/livepeer/ai-runner/blob/145f5ca7a9a2ab63c549764a42d4bdc4e815a8b3/runner/app/live/streamer/process.py#L170-L191